File: test/language/functions/java/functions/structure_array/StructureArrayTest.java
package functions.structure_array;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;
import java.util.Arrays;

import org.junit.jupiter.api.Test;

import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

public class StructureArrayTest
{
    @Test
    public void checkStructureArrayElement0() throws IOException
    {
        checkStructureArray(0);
    }

    @Test
    public void checkStructureArrayElement1() throws IOException
    {
        checkStructureArray(1);
    }

    @Test
    public void checkStructureArrayElement2() throws IOException
    {
        checkStructureArray(2);
    }

    private byte[] writeStructureArrayToByteArray(int pos) throws IOException
    {
        ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        writer.writeBits(NUM_ITEM_ELEMENTS, 16);

        for (Item item : ITEMS)
        {
            writer.writeBits(item.getA(), 8);
            writer.writeBits(item.getB(), 8);
        }

        writer.writeBits(pos, 16);

        writer.close();

        return writer.toByteArray();
    }

    private StructureArray createStructureArray(int pos)
    {
        final StructureArray structureArray = new StructureArray();

        structureArray.setNumElements(NUM_ITEM_ELEMENTS);
        structureArray.setValues(ITEMS);

        structureArray.setPos(pos);

        return structureArray;
    }

    private void checkStructureArray(int pos) throws IOException
    {
        final StructureArray structureArray = createStructureArray(pos);
        assertEquals(ITEMS[pos], structureArray.funcGetElement());

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        structureArray.write(writer);
        final byte[] writtenByteArray = writer.toByteArray();

        final byte[] expectedByteArray = writeStructureArrayToByteArray(pos);
        assertTrue(Arrays.equals(expectedByteArray, writtenByteArray));

        final ByteArrayBitStreamReader reader =
                new ByteArrayBitStreamReader(writtenByteArray, writer.getBitPosition());
        final StructureArray readStructureArray = new StructureArray(reader);
        assertEquals(structureArray, readStructureArray);
    }

    private static final Item[] ITEMS = new Item[] {
            new Item((short)1, (short)2), new Item((short)3, (short)4), new Item((short)5, (short)6)};
    private static final int NUM_ITEM_ELEMENTS = ITEMS.length;
}

File: test/extensions/language/functions/python/StructureArrayTest.py
import zserio

import Functions


class StructureArrayTest(Functions.TestCase):
    @classmethod
    def setUpClass(cls):
        super(StructureArrayTest, cls).setUpClass()

        cls.ITEMS = [cls.api.Item(1, 2), cls.api.Item(3, 4), cls.api.Item(5, 6)]
        cls.NUM_ITEMS = len(cls.ITEMS)

    def testStructureArrayElement0(self):
        self._checkStructureArray(0)

    def testStructureArrayElement1(self):
        self._checkStructureArray(1)

    def testStructureArrayElement2(self):
        self._checkStructureArray(2)

    def _writeStructureArrayToStream(self, writer, pos):
        writer.write_bits(self.NUM_ITEMS, 16)

        for item in self.ITEMS:
            writer.write_bits(item.value_a, 8)
            writer.write_bits(item.value_b, 8)

        writer.write_bits(pos, 16)

    def _createStructureArray(self, pos):
        return self.api.StructureArray(self.NUM_ITEMS, self.ITEMS, pos)

    def _checkStructureArray(self, pos):
        structureArray = self._createStructureArray(pos)
        self.assertEqual(self.ITEMS[pos], structureArray.get_element())

        writer = zserio.BitStreamWriter()
        structureArray.write(writer)
        expectedWriter = zserio.BitStreamWriter()
        self._writeStructureArrayToStream(expectedWriter, pos)
        self.assertTrue(expectedWriter.byte_array == writer.byte_array)
        self.assertTrue(expectedWriter.bitposition == writer.bitposition)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readStructureArray = self.api.StructureArray.from_reader(reader)
        self.assertEqual(structureArray, readStructureArray)
